apps/bluetooth/js/modules/transfer_manager.js

```diff
@@ -0,0 +1,372 @@
+           activityRequest.source.data.blobs.length > 0)) {
+        this._activity = activityRequest;
+        // Source info is enought to servie file transfer.
+        // Start to init and observe Bluetooth relative modules.
+        this._observeDefaultAdapter();
+        this._observeBluetoothEnabled();
+        this._watchEventFromDevicePicker();
+        DevicePickerPanel.visible = true;
+      } else {
+        Debug('_activityHandler(): Cannot transfer without blobs data!');
+        this._cannotTransfer();
+      }
+    },
+
+    /**
+     * Show alert dialog. End the activity service while a user confirm to close
+     * the dialog.
+     *
+     * @access private
+     * @memberOf TransferManager
+     */
+    _cannotTransfer: function bttm__cannotTransfer() {
+      Debug('_cannotTransfer(): show alert dialog!!');
+      // show an alert with the overdue message
+      if (!CannotTransferDialog.isVisible) {
+        CannotTransferDialog.showConfirm(() => {
+          // Post error with reason then end the activity service.
+          this._endTransferWithReason('cancelled');
+        });
+      }
+    },
+
+    /**
+     * Post result/error to end the activity service.
+     *
+     * @access private
+     * @memberOf TransferManager
+     */
+    _endTransferWithReason: function bttm__endTransferWithReason(reason) {
+      Debug('_endTransferWithReason(): reason = ' + reason);
+      if (reason === 'transferred') {
+        this._activity.postResult(reason);
+      } else if (reason === 'cancelled') {
+        this._activity.postError(reason);
+      }
+      
+      this._activity = null;
+    },
+
+    /**
+     * Observe and get default adapter from bluetooth adapter manager.
+     *
+     * @access private
+     * @memberOf TransferManager
+     */
+    _observeDefaultAdapter: function bttm__observeDefaultAdapter() {
+      // Observe 'defaultAdapter' property for reaching default adapter.
+      AdapterManager.observe('defaultAdapter',
+        this._onDefaultAdapterChanged.bind(this));
+      this._onDefaultAdapterChanged(AdapterManager.defaultAdapter);
+    },
+
+    /**
+     * Observe and get 'enabled' property from bluetooth context.
+     *
+     * @access private
+     * @memberOf TransferManager
+     */
+    _observeBluetoothEnabled: function bttm__observeBluetoothEnabled() {
+      // Observe Bluetooth 'enabled' property from hardware side.
+      BtContext.observe('enabled', this._onBluetoothEnabledChanged.bind(this));
+      this._onBluetoothEnabledChanged(BtContext.enabled);
+    },
+
+    /**
+     * 'defaultAdapter' change event handler from adapter manager for
+     * updating it immediately.
+     *
+     * @access private
+     * @memberOf TransferManager
+     * @param {Object BluetoothAdapter} newAdapter
+     */
+    _onDefaultAdapterChanged: 
+    function bttm__onDefaultAdapterChanged(newAdapter, oldAdapter) {
+      if ((oldAdapter !== null) && (newAdapter === null)) {
```

Comment:
Okay, I forgot that when bluetooth is disabled we can still get the default adapter.